File: wasm-of-ocaml/benchmarks/C/nbody_100.c
#include <math.h>

const double pi = 3.141592653589793;
const double solar_mass = 4.0 * pi * pi;
const double days_per_year = 365.24;

// Should really pay closure attention to use of const/non-const in other C programs, affects optimisations
struct planet {
    float x, y, z, vx, vy, vz;
    const float mass;
};
typedef struct planet Planet;

void advance(Planet **bodies, int num_bodies, double dt){
    for (int i = 0; i < num_bodies; i++){
        Planet *b = bodies[i];
        for (int j = i + 1; j < num_bodies; j++){
            Planet *b2 = bodies[j];
            double dx = b->x - b2->x, dy = b->y - b2->y, dz = b->z - b2->z;
            double dist2 = dx * dx + dy * dy + dz * dz;
            double mag = dt/(dist2 * sqrt(dist2));

            b->vx -= dx * b2->mass * mag;
            b->vy -= dy * b2->mass * mag;
            b->vz -= dz * b2->mass * mag;

            b2->vx += dx * b->mass * mag;
            b2->vy += dy * b->mass * mag;
            b2->vz += dz * b->mass * mag;
        }
    }
    for (int i = 0; i < num_bodies; i++){
        Planet *b = bodies[i];
        b->x += dt * b->vx;
        b->y += dt * b->vy;
        b->z += dt * b->vz;
    }
}

double energy(Planet **bodies, int num_bodies){
  double e = 0;
  for (int i = 0; i < num_bodies; i++){
    Planet *b = bodies[i];
    e += 0.5 * b->mass * (b->vx * b->vx + b->vy * b->vy + b->vz * b->vz);
    for (int j = i + 1; j < num_bodies; j++){
      Planet *b2 = bodies[j];
      double dx = b->x - b2->x, dy = b->y - b2->y, dz = b->z - b2->z;
      double distance = sqrt(dx * dx + dy * dy + dz * dz);
      e -= (b->mass * b2->mass) / distance;
    }
  }
  return e;
}

void offset_momentum(Planet **bodies, int num_bodies){
  double px = 0, py = 0, pz = 0;
  for (int i = 0; i < num_bodies; i++){
    px += bodies[i]->vx * bodies[i]->mass;
    py += bodies[i]->vy * bodies[i]->mass;
    pz += bodies[i]->vz * bodies[i]->mass;
  }
  bodies[0]->vx = - px / solar_mass;
  bodies[0]->vy = - py / solar_mass;
  bodies[0]->vz = - pz / solar_mass;
}

Planet jupiter = {4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
    1.66007664274403694e-03 * days_per_year, 7.69901118419740425e-03 * days_per_year,
    -6.90460016972063023e-05 * days_per_year, 9.54791938424326609e-04 * solar_mass};

Planet saturn = {  8.34336671824457987e+00,
                4.12479856412430479e+00,
                -4.03523417114321381e-01,
                -2.76742510726862411e-03 * days_per_year,
                4.99852801234917238e-03 * days_per_year,
                2.30417297573763929e-05 * days_per_year,
               2.85885980666130812e-04 * solar_mass    };

Planet uranus = {  1.28943695621391310e+01,
                -1.51111514016986312e+01,
                -2.23307578892655734e-01,
                2.96460137564761618e-03 * days_per_year,
                2.37847173959480950e-03 * days_per_year,
                -2.96589568540237556e-05 * days_per_year,
               4.36624404335156298e-05 * solar_mass     };

Planet neptune = {  1.53796971148509165e+01,
                 -2.59193146099879641e+01,
                 1.79258772950371181e-01,
                 2.68067772490389322e-03 * days_per_year,
                 1.62824170038242295e-03 * days_per_year,
                 -9.51592254519715870e-05 * days_per_year,
                5.15138902046611451e-05 * solar_mass   };

Planet sun = {  0.,   0.,   0.,   0.,   0.,  0., solar_mass };

int main(){
    Planet *bodies[] = {&sun, &jupiter, &saturn, &uranus, &neptune};
    int num_bodies = 5;
    int n = 100;
    double e = 0.;
    offset_momentum(bodies, num_bodies);
    e = energy(bodies, num_bodies);
    for (int i = 0; i <= n; i++){
        advance(bodies, num_bodies, 0.01);
    }
    e += energy(bodies, num_bodies);
    // to stop it being optimised away
    return (int)e;
}

File: wasm-of-ocaml/benchmarks/Ocaml/nbody_100.ml
(* The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 Copyright © 2004-2008 Brent Fulgham, 2005-2021 Isaac Gouy
 All rights reserved.
 Redistribution and use in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation and/or other materials provided
    with the distribution.

 3. Neither the name "The Computer Language Benchmarks Game" nor the name "The Benchmarks Game" nor the name
    "The Computer Language Shootout Benchmarks" nor the names of its contributors may be used to endorse or
     promote products derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 * Contributed by Troestler Christophe
 *)

let pi = 3.141592653589793
let solar_mass = 4. *. pi *. pi
let days_per_year = 365.24

type planet = { mutable x : float; mutable y : float;  mutable z : float;
                mutable vx: float;  mutable vy: float;  mutable vz: float;
                mass : float }

let advance bodies num_bodies dt =
  let n = num_bodies - 1 in
  for i = 0 to n do
    let b = bodies.(i) in
    for j = i+1 to n do
      let b' = bodies.(j) in
      let dx = b.x -. b'.x  and dy = b.y -. b'.y  and dz = b.z -. b'.z in
      let dist2 = dx *. dx +. dy *. dy +. dz *. dz in
      let mag = dt /. (dist2 *. sqrt(dist2)) in

      b.vx <- b.vx -. dx *. b'.mass *. mag;
      b.vy <- b.vy -. dy *. b'.mass *. mag;
      b.vz <- b.vz -. dz *. b'.mass *. mag;

      b'.vx <- b'.vx +. dx *. b.mass *. mag;
      b'.vy <- b'.vy +. dy *. b.mass *. mag;
      b'.vz <- b'.vz +. dz *. b.mass *. mag;
    done
  done;
  for i = 0 to n do
    let b = bodies.(i) in
    b.x <- b.x +. dt *. b.vx;
    b.y <- b.y +. dt *. b.vy;
    b.z <- b.z +. dt *. b.vz;
  done

let energy bodies num_bodies =
  let e = ref 0. in
  for i = 0 to num_bodies - 1 do
    let b = bodies.(i) in
    e := (!e) +. 0.5 *. b.mass *. (b.vx *. b.vx +. b.vy *. b.vy +. b.vz *. b.vz);
    for j = i+1 to num_bodies - 1 do
      let b' = bodies.(j) in
      let dx = b.x -. b'.x  and dy = b.y -. b'.y  and dz = b.z -. b'.z in
      let distance = sqrt(dx *. dx +. dy *. dy +. dz *. dz) in
      e := (!e) -. (b.mass *. b'.mass) /. distance
    done
  done;
  !e

let offset_momentum bodies num_bodies =
   let px = ref 0. and py = ref 0. and pz = ref 0. in
  for i = 0 to num_bodies - 1 do
    px := (!px) +. bodies.(i).vx *. bodies.(i).mass;
    py := (!py) +. bodies.(i).vy *. bodies.(i).mass;
    pz := (!pz) +. bodies.(i).vz *. bodies.(i).mass;
  done;
  bodies.(0).vx <- -. (!px) /. solar_mass;
  bodies.(0).vy <- -. (!py) /. solar_mass;
  bodies.(0).vz <- -. (!pz) /. solar_mass

let jupiter = { x = 4.84143144246472090e+00;
                y = -1.16032004402742839e+00;
                z = -1.03622044471123109e-01;
                vx = 1.66007664274403694e-03 *. days_per_year;
                vy = 7.69901118419740425e-03 *. days_per_year;
                vz = -6.90460016972063023e-05 *. days_per_year;
                mass = 9.54791938424326609e-04 *. solar_mass;    }

let saturn = { x = 8.34336671824457987e+00;
               y = 4.12479856412430479e+00;
               z = -4.03523417114321381e-01;
               vx = -2.76742510726862411e-03 *. days_per_year;
               vy = 4.99852801234917238e-03 *. days_per_year;
               vz = 2.30417297573763929e-05 *. days_per_year;
               mass = 2.85885980666130812e-04 *. solar_mass;     }

let uranus = { x = 1.28943695621391310e+01;
               y = -1.51111514016986312e+01;
               z = -2.23307578892655734e-01;
               vx = 2.96460137564761618e-03 *. days_per_year;
               vy = 2.37847173959480950e-03 *. days_per_year;
               vz = -2.96589568540237556e-05 *. days_per_year;
               mass = 4.36624404335156298e-05 *. solar_mass;     }

let neptune = { x = 1.53796971148509165e+01;
                y = -2.59193146099879641e+01;
                z = 1.79258772950371181e-01;
                vx = 2.68067772490389322e-03 *. days_per_year;
                vy = 1.62824170038242295e-03 *. days_per_year;
                vz = -9.51592254519715870e-05 *. days_per_year;
                mass = 5.15138902046611451e-05 *. solar_mass;   }

let sun = { x = 0.;  y = 0.;  z = 0.;  vx = 0.;  vy = 0.; vz = 0.;
            mass = solar_mass; }

let bodies = [| sun; jupiter; saturn; uranus; neptune |]
(* Needed since Array.Length not implemented *)
let num_bodies = 5

let f =
  (* Number of iterations *)
  let n = 100 in
  (* Use result, since can't print it *)
  let e = ref 0. in
  offset_momentum bodies num_bodies;
  e := energy bodies num_bodies;
  for i = 1 to n do advance bodies num_bodies 0.01 done;
  e := energy bodies num_bodies;
  !e
